File: proj.py
#! /home/dustin/tmp/ve/bin/python

## TODO: rewrite in Go

import sys
import os
import yaml
import json
import tempfile
import argparse

SUPPORTED_SHELLS = ('bash',)
PROJ_PATH = os.path.abspath(__file__)

def msg(message, fatal=False):
    print(message)
    if fatal:
        sys.exit(1)

## shell


def activate_virtualenv(config, context, ctx_value, env, args):
    ve = os.path.abspath(ctx_value)
    return 'source "{}/bin/activate"'.format(ve)

context_handlers = {
    'activate_virtualenv': activate_virtualenv,
}


def do_shell(config, context):
    msg("starting shell")

    rc_fd, rc_filename = tempfile.mkstemp()

    try:
        args = {
            'bash': ['bash', '--rcfile', rc_filename, '-i'],
        }[context['shell']]

        rc_content = []

        # run the user's shell config first
        rc_content.append({
            'bash': '[ -f ~/.bashrc ] && . ~/.bashrc',
        }[context['shell']])

        env = os.environ.copy()
        for c in context['modifiers']:
            name, ctx_value = c.items()[0]
            msg("applying {}: {!r}".format(name, ctx_value))
            rc = context_handlers[name](config, context, ctx_value, env, args)
            if rc:
                rc_content.append(rc)

        # delete the rc temp file
        rc_content.append({
            'bash': 'rm -f "{}"'.format(rc_filename),
        }[context['shell']])

        # write out the rcfile content and close it
        os.fdopen(rc_fd, "w").write("\n\n".join(rc_content))

        # exec the shell
        os.execvpe(args[0], args, env)
    except Exception:
        os.unlink(rc_filename)
        raise


def _local_reexec(context, elt, path, env_config=None):
    # re-run proj within the new environment
    args = [sys.executable, PROJ_PATH]

    # fork off a process to write the config to the process we exec
    r, w = os.pipe()
    if os.fork() == 0:
        os.close(r)
        json.dump(context, os.fdopen(w, "w"))
        os._exit(0)

    os.close(w)
    args.extend(['--cfd', str(r)])
    if env_config:
        args.extend(['--env-config', env_config])
    args.append(path)
    os.execv(sys.executable, args)

## context config


def ctx_shell(context, name, args):
    shell = args
    if shell not in SUPPORTED_SHELLS:
        msg("** shell {!r} not supported".format(shell), fatal=True)
    context['shell'] = shell

def ctx_default(context, name, args):
    context['modifiers'].append({name: args})

context_config_handlers = {
    'shell': ctx_shell,
}

def update_context(context, ctx_cfg):
    if not isinstance(ctx_cfg, dict) or len(ctx_cfg) != 1:
        msg("** invalid context config", fatal=True)

    name, args = ctx_cfg.items()[0]
    fn = context_config_handlers.get(name, ctx_default)
    return fn(context, name, args)

## traversal


def child_cd(config, context, elt, path, arg):
    if not isinstance(arg, dict):
        arg = {'dir': arg}
    if 'dir' not in arg:
        msg("** no directory specified for 'cd' in {}".format(config['__file__']), fatal=True)
    env_config = None
    if arg.get('config'):
        env_config = os.path.abspath(arg['config'])
    os.chdir(arg['dir'])
    _local_reexec(context, elt, path, env_config=env_config)


child_methods = {
    'cd': child_cd,
}

def start_child(config, context, elt, path):
    msg("entering " + elt)
    cfg_file = config['__file__']

    if elt not in config['children']:
        msg("** no such child project '{}' in {}".format(elt, cfg_file), fatal=True)
        return

    child = config['children'][elt]
    if len(child) != 1:
        msg("** malformed child project '{}' in {}".format(elt, cfg_file), fatal=True)
    child_method, method_arg = child.items()[0]

    if child_method not in child_methods:
        msg("** no such child method '{}' in {}".format(child_method, cfg_file), fatal=True)

    child_methods[child_method](config, context, elt, path, method_arg)

## config management


def get_config(args):
    fn = os.path.expanduser('~/.projrc.yml')
    if os.path.exists(fn):
        rc_config = yaml.load(open(fn))
    else:
        rc_config = {}

    dirname = os.path.basename(os.path.abspath('.'))
    print args.env_config
    if args.env_config:
        if not os.path.exists(args.env_config):
            msg("** explicit config file {!r} not found".format(os.path.abspath(args.env_config)), fatal=True)
        configs = [args.env_config]
    else:
        configs = ['.proj.yml', os.path.join('..', "{}-proj.yml".format(dirname))]
    for filename in configs:
        if os.path.exists(filename):
            env_config = yaml.load(open(filename))
            break
    else:
        env_config = {}
        filename = os.path.abspath('.')
    msg("loading env config from {!r}".format(os.path.abspath(filename)))

    rc_config.update({
        '__file__': filename,
        'children': env_config.get('children', {}),
        'context': env_config.get('context', [])
    })

    return rc_config


def load_context(args):
    if args.cfd:
        return json.load(os.fdopen(args.cfd))
    else:
        return {
            'shell': SUPPORTED_SHELLS[0],
            'modifiers': [],
        }


def main(argv):
    try:
        parser = argparse.ArgumentParser(description='Set up shell environments for projects')
        parser.add_argument('--cfd', type=int)
        parser.add_argument('path')
        parser.add_argument('--env-config')

        args = parser.parse_args(argv[1:])

        config = get_config(args)

        context = load_context(args)
        for ctx_row in config['context']:
            update_context(context, ctx_row)
        
        if args.path == '':
            do_shell(config, context)
        else:
            if '/' in args.path:
                elt, path = args.path.split('/', 1)
            else:
                elt, path = args.path, ''
            start_child(config, context, elt, path)
    except SystemExit:
        pass

    # if we end up here, something went wrong, so hang around long enough to see the message
    raw_input("ENTER to exit ")


if __name__ == "__main__":
    main(sys.argv)

File: proj.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"github.com/kylelemons/go-gypsy/yaml"
	"log"
	"os"
	"path"
	"strings"
	"syscall" // TODO: don't use
)

/* Config handling */

type ConfigElt struct {
	Type string
	Args yaml.Node
}

type Config struct {
	Filename string
	Children map[string]ConfigElt
	Contexts []ConfigElt

	raw yaml.Node
}

func (c Config) String() string {
	return fmt.Sprintf("{\n    Filename: %#v\n    Children: %#v\n    Contexts: %#v\n}",
		c.Filename, c.Children, c.Contexts)
}

func load_config(env_config string) Config {
	var config Config
	var filename string

	if len(env_config) > 0 {
		filename = env_config
	} else {
		wd, err := os.Getwd()
		if err != nil {
			log.Panic(err)
		}
		filename = path.Clean(path.Join(wd, ".proj.yml"))
		if _, err := os.Stat(filename); err != nil {
			log.Printf("fallback, %#v", filename)
			dirname := path.Base(wd)
			filename = path.Clean(path.Join(wd, fmt.Sprintf("../%s-proj.yml", dirname)))
		}
	}

	if _, err := os.Stat(filename); err != nil {
		log.Panic(fmt.Sprintf("Config file '%s' not found", filename))
	}
	config.Filename = filename

	// TODO: load ~/.projrc.yml too
	file, err := yaml.ReadFile(config.Filename)
	if err != nil {
		log.Panic(err)
	}
	config.raw = file.Root

	// parse children
	config.Children = make(map[string]ConfigElt)
	child_node, err := yaml.Child(config.raw, ".children")
	if err == nil {
		for name, value := range child_node.(yaml.Map) {
			val_map := value.(yaml.Map)
			if len(val_map) != 1 {
				log.Panic("malformed configuration 1")
			}
			for typ, args := range val_map {
				config.Children[name] = ConfigElt{
					Type: typ,
					Args: args}
			}
		}
	}

	// parse contexts
	contexts_node, err := yaml.Child(config.raw, ".context")
	if err == nil {
		contexts_list := contexts_node.(yaml.List)
		config.Contexts = make([]ConfigElt, len(contexts_list))
		for i, ctx := range contexts_list {
			ctx_map := ctx.(yaml.Map)
			if len(ctx_map) != 1 {
				log.Panic("malformed configuration 2")
			}
			for typ, args := range ctx_map {
				config.Contexts[i] = ConfigElt{
					Type: typ,
					Args: args}
			}
		}
	} else {
		config.Contexts = make([]ConfigElt, 0)
	}

	return config
}

/* Context handling */

type Context struct {
	Shell     string
	Path      []string
	Modifiers []interface{}
}

func load_context(cfd int) Context {
	if cfd == 0 {
		return Context{
			Shell:     "bash", // TODO from supported_shells
			Path:      []string{},
			Modifiers: make([]interface{}, 0),
		}
	}

	// read from the given file descriptor and close it
	ctxfile := os.NewFile(uintptr(cfd), "ctxfile")
	decoder := json.NewDecoder(ctxfile)
	context := Context{}
	err := decoder.Decode(&context)
	if err != nil {
		log.Panic(err)
	}
	ctxfile.Close()

	log.Printf("got context %#v\n", context)

	return context
}

func write_context(context Context, w *os.File) {
	encoder := json.NewEncoder(w)
	err := encoder.Encode(context)
	if err != nil {
		log.Panic(err)
	}
	w.Close()
}

/* shell handling */

func do_shell(config Config, context Context) {
	log.Printf("do_shell(%+v, %+v)\n", config, context) // TODO

	if context.Shell != "bash" {
		log.Fatalf("unkonwn shell %s", context.Shell)
	}

	// TODO: search PATH for the shell
	// TODO: execute a shell script
	err := syscall.Exec("/usr/bin/bash", []string{"bash"}, nil)
	log.Panic(err)
}

/* child handling */

func local_reexec(context Context, path string) {
	log.Printf("running %s", os.Args[0])

	// Fork a new child, then write out the context and exit.  This results
	// in a rapid cascade of sub-processes, with ppid=1, but Go doesn't allow
	// raw forks, it seems. (TODO)
	r, w, err := os.Pipe()
	if err != nil {
		log.Panic(err)
	}

	args := []string{os.Args[0], "--cfd", "3", path}
	procattr := os.ProcAttr{
		Files: []*os.File{os.Stdin, os.Stdout, os.Stderr, r},
	}
	proc, err := os.StartProcess(args[0], args, &procattr)
	if err != nil {
		log.Panic(err)
	}

	err = r.Close()
	if err != nil {
		log.Panic(err)
	}

	write_context(context, w)

	// TODO: would rather just exit here, but then the caller forgets about us
	proc.Wait()
	os.Exit(0)
}

// utility function to get a child of a YAML node, or if the YAML node is not a
// Map, assume that it is the expected value.  This allows things like "cd:
// somedir" as a shorthand for "cd: dir: somedir".
func default_child(args yaml.Node, key string) (yaml.Node, error) {
	_, ok := args.(yaml.Map)
	if !ok {
		return args, nil
	} else {
		return yaml.Child(args, key)
	}
}

func start_child_cd(config Config, context Context, child ConfigElt, path string) {
	// TODO: handle Args being a string
	dir_node, err := default_child(child.Args, "dir")
	if err != nil {
		log.Panic(err)
	}
	// TODO: handle not a scalar
	dir_scalar, ok := dir_node.(yaml.Scalar)
	if !ok {
		log.Panic("invalid directory %#v", dir_node)
	}
	dir := dir_scalar.String()

	// TODO: handle 'config' option, sending --env-config

	err = os.Chdir(dir)
	if err != nil {
		log.Panic(err)
	}

	local_reexec(context, path)
}

func start_child(config Config, context Context, elt string, path string) {
	log.Printf("start_child(%+v, %+v, %+v, %+v)\n", config, context, elt, path)
	child, ok := config.Children[elt]
	if !ok {
		log.Fatalf("No such child %s", elt)
	}

	// add the element to the context to be handed to the child
	context.Path = append(context.Path, elt)

	if child.Type == "cd" {
		start_child_cd(config, context, child, path)
	} else {
		log.Panic("unknown child or child type")
	}
}

/* main */

func run(cfd int, env_config string, path string) {
	config := load_config(env_config)
	context := load_context(cfd)

	// TODO: update context based on config

	// either start a shell or enter the next path element
	if len(path) == 0 {
		do_shell(config, context)
	} else {
		i := strings.Index(path, "/")
		if i < 0 {
			start_child(config, context, path, "")
		} else {
			start_child(config, context, path[:i], path[i+1:])
		}
	}
}

func main() {
	cfd := flag.Int("cfd", 0, "(internal use only)")
	env_config := flag.String("env-config", "", "(internal use only)")

	flag.Parse()

	args := flag.Args()
	if len(args) != 1 {
		log.Panic("Path is required")
	}

	path := args[0]

	run(*cfd, *env_config, path)
}
